### PR TITLE
Enforce signatures on post/patch operations

### LINKED ISSUE #44: Sign messages on challenge / challenge review submission
For extra security, we should:

- make the user sign a when submitting a challenge.
- make the admin sign when submitting a challenge submission.

Rule of thumb: make signing mandatory when writing the database.

### PR DESCRIPTION
**Pull request by carletex**
Fixes #44 

I worked on this with @dgrcode and I've just gave it the final touches.

This PR enforces the use of signatures on challenge /  challenge review submissions. Not sure if we are missing something with the signatures, but I think this should be secure enough.

Let me know what you think.

### CODE REVIEW

**Review by dgrcode — APPROVED** [review]
I'll address all the comments besides the middleware and commit. I'll also open a new issue for the middleware improvement.

File: packages/backend/index.js [hunk 1]
```diff
@@ -71,34 +73,36 @@ app.get("/builders/:builderAddress", async (req, res) => {
...
-
-      const jwt = await firebaseAdmin.auth().createCustomToken(recovered, { isAdmin: !!userObject.isAdmin });
-
-      response.json({ ...userObject, token: jwt });
-    } else {
-      response.status(401).send(" 🚫 Signature verification failed! Please reload and try again. Sorry! 😅");
-    }
+    // Retrieve existing user.
+    console.log("Retrieving existing user: ", userAddress);
+    userObject = user.data();
```
[the remark this answers is not in the document]
Comment (sogasg) [reply packages/backend/index.js:100]: The only things that are used on sign-in are the JWT and isAdmin from the user object. We should probably rewrite that part to get the isAdmin variable from the JWT claims.

[the remark this answers is not in the document]
Comment (dgrcode) [reply packages/backend/index.js:100]: yep, that sounds in line with #50. I'm going to push a commit where I removed the backend side of this, but I'm still sending isAdmin and letting the front end work as it was working until now. #50 should take care of the front end to read stuff from the JWT

File: packages/backend/utils/sign.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  switch (messageId) {
+    case "challengeSubmit":
+      // ToDo. Submission nonce.
+      return JSON.stringify({
+        messageId,
+        address: options.address,
+        challengeId: options.challengeId,
+      });
+    case "challengeReview":
+      // ToDo. Submission nonce.
```
[the remark this answers is not in the document]
Comment (dgrcode) [reply packages/backend/utils/sign.js:13]: Moved this comment to #30. This should be handled there as well.

File: packages/backend/index.js [hunk 3]
```diff
@@ -116,10 +120,21 @@ app.get("/user", userOnly, async (request, response) => {
...
+  const verifyOptions = {
+    messageId: "challengeSubmit",
+    address,
+    challengeId,
+  };
+
+  if (!verifySignature(signature, verifyOptions)) {
+    response.status(401).send(" 🚫 Signature verification failed! Please reload and try again. Sorry! 😅");
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (dgrcode) [reply packages/backend/index.js:131]: Moved to #51

File: packages/backend/utils/sign.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+  }
+};
+
+const verifySignature = (signature, verifyOptions) => {
+  const trustedMessage = getSignMessageForId(verifyOptions.messageId, verifyOptions);
+  const signingAddress = ethers.utils.verifyMessage(trustedMessage, signature);
+
+  console.log("trustedMessage", trustedMessage);
+  console.log("signingAddress       ", signingAddress);
+  console.log("verifyOptions.address", verifyOptions.address);
```
[the remark this answers is not in the document]
Comment (dgrcode) [reply packages/backend/utils/sign.js:35]: Moved to #52

**Comment by dgrcode** [comment]
I made a few changes related to the comments I added before. Just merged into master a few hours before dogfooding, which is probably not recommendable, although it'd be awesome to show this feature today.

@sogasg Could you please give it a run in your local before dogfooding? I checked everything works on my machine, but this things are mysterious some times, especially on live demos.

In case things blow up, the commit master had as HEAD before merging was https://github.com/moonshotcollective/scaffold-directory/commit/2af7e33b8ba20c1d049a02f71c051a9b6c447ce8. Feel free to
```git reset --hard 2af7e33b8ba20c1d049a02f71c051a9b6c447ce8```
if needed. I hope this is not necessary! 🤞

**Comment by carletex** [comment]
Thanks @dgrcode !!

Just tested locally and everything seems to work.

@sogasg Good luck at the dogfooding, let us know how it goes.

**Comment by sogasg** [comment]
Awesome work! 

Everything is working locally for me too :D

**Pull request merged**